### ISSUE TITLE
Feature Request: Desktop Notifications

### ISSUE THREAD

**Issue by jwhipp** [microsoft/WSL#2466]
>Microsoft Windows [Version 10.0.15063]

Desktop notifications from Linux apps don't appear in Windows, it would be nice if this could work.

Example test:

$ sudo apt-get install libnotify-bin
$ notify-send "TEST"

Feature request is to have the notification interface to allow the action center to receive notifications.

**Comment by shevegen** [comment]
I agree. It probably does not have a huge priority presently though.

There will be a larger update towards the middle of October so perhaps afterwards some lower priority things may receive more attention. :)

**Comment by sunilmut** [comment]
Thanks for the idea. Marking it so. Please open a ticket at the [WSL user voice](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo) for any feature requests. It helps us prioritize for future.

**Comment by crozone** [comment]
libnotify uses the [Desktop Notifications Specification](https://developer.gnome.org/notification-spec/), which works on top of D-Bus.

This means that the solution can probably already be accomplished by writing a Windows-land daemon that connects to the Linux-land D-Bus daemon over TCP, and creates Windows notification popups.

I might even give this a go in C# if it's not too hard.

**Comment by sunilmut** [comment]
@crozone - Interesting. I am just trying to visualize how this would work. So, the `notify-send` (or any other command) would go over `libnotify` to the WSL D-Bus service, which can publish the message over to the `Desktop Notification Specification` interface. And, if there is a Win32 Desktop Notification server that can consume these messages, then it can create the necessary popups?

**Comment by crozone** [comment]
@sunilmut sorry for the massively delayed reply, but yes.

The trick is that the win32 application can connect to the Linux D-Bus using a TCP connection to localhost. In theory, it would make the most sense for an X Window Sever (like XMing) to implement it on the win32 side, since IIRC it already establishes a connection to a D-Bus channel relating to that desktop. However, failing that, a stand alone application can work, but I'm not 100% sure how it would find the right D-Bus channel.

**Comment by therealkenc** [comment]
> it would make the most sense for an X Window Sever (like XMing) to implement it on the win32 side

Wrong protocol. All the `dbus` action happens on the WSL side; an X Server (be it XMing, VcXsrv, native Linux Xorg) has no knowledge of `dbus` or what is being displayed in a popup window. It speaks X11 protocol.

The guys who obscure half their screen with Ubuntu wallpaper over in #637 already have desktop notifications. `notify-send` works fine for them; or at least, it should. I think you can fire up [`unity-notifications`](https://launchpad.net/unity-notifications) alone and get exactly what you describe (ignoring that `unity-notifications` probably falls on its face without Unity around). The `xfce4` guys have something similar.

If **all** you want is CLI `notify-send`, actual Win32 Desktop Notifications, and no X11 to speak of, you could write your own Desktop Notification Specification service like Sunil describes. But the `dbus` session server and the notification client _still live on the WSL side_, communicating over an AF_UNIX socket. Typically you'd talk to the service [using ssh](http://gabriel.sourceforge.net/howto.html). 

Basically, on the Windows side you write an app that remote launches (on WSL) some Python script that sits there waiting for Desktop Notification Specification messages to arrive, and spits the notification text out to `stdout` in some formatted manner. Then on the windows side when new message text arrives over ssh you push a Windows Desktop notification.

I mean, in a fever dream kind of way.

**Comment by crozone** [comment]
>Wrong protocol. All the dbus action happens on the WSL side; an X Server (be it XMing, VcXsrv, native Linux Xorg) has no knowledge of dbus or what is being displayed in a popup window. It speaks X11 protocol.

Ahh, thanks. I thought there was some magic there that would allow X apps to "find" the D-Bus channel for that session, but maybe that's something else.

> Basically, on the Windows side you write an app that remote launches (on WSL) some Python script that sits there waiting for Desktop Notification Specification messages to arrive, and spits the notification text out to stdout in some formatted manner. Then on the windows side when new message text arrives over ssh you push a Windows Desktop notification.

I'm pretty sure you don't need to do this though, you can get D-Bus to listen for TCP connections, instead of just the AF_UNIX socket. This would let the Windows-side client speak the notification protocol natively.

**Comment by therealkenc** [comment]
> you can get D-Bus to listen for TCP connections

Yeah I almost mentioned that but didn't want to yammer (knowing in my heart someone would bring it up anyway). That is technically correct, but would mean the mother-of-all dependency inversions. In other words, installing a hypothetical "WSL desktop notify package" would mean forcing TCP on _everything else_. And no one would do that. This is why `dbus-broker` just scrapped the idea entirely (search "Local Only" in the article).

**Comment by DarthSpock** [comment]
@therealkenc @sunilmut This can probably be looked at now using AF_Unix interop. Not sure how simple it would be but it would be nice to have this small but useful feature.

**Comment by crozone** [comment]
I was just thinking about how the new AF_UNIX interop stuff could fit into this. As far as I understand it, a socket can only be consumed from Win32 land if it is created within the path of a Win32 accessible filesystem, and can then only be accessed by the creator (on the UNIX side), and Win32 apps.

Is D-Bus able to create two unix sockets and listen on both? (one for the unix side in /tmp, and one on the Win32 side, somewhere else)

**Comment by therealkenc** [comment]
Or the other direction; even easier.

In Windows, launch Powershell as Administrator and do:
```
PS C:\WINDOWS\system32> Install-Module -Name BurntToast
```
On WSL:
```
$ alias toast='powershell.exe -command New-BurntToastNotification'
```
And finally:
```
$ toast -Text hello
```
FTFY

**Comment by therealkenc** [comment]
So a follow-up on this, just because I took the time-sink to do a "what would it take" look. I hacked up [`notification-thing`](https://github.com/mk-fg/notification-thing) to spawn New-BurntToastNotification instead of popping up a GTK3+ window. It even has a [liberal license](https://github.com/mk-fg/notification-thing/blob/master/COPYING). We're talking trivial changes here. That gets you a working `notify-send` from `libnotify-bin` as asked for in the OP.

Long story short (I've deleted a third of this post), The Desktop Notifications Specification operates in a **session**. This isn't related to the `<listen>` option in `/usr/share/dbus-1/system.conf` (`AF_UNIX` or otherwise). You _can_ in theory and practice create a remote login dbus session for every WSL terminal, which goes to Ben's comment in #816 [(here)](https://github.com/Microsoft/WSL/issues/816#issuecomment-278693987). But the end result is you're just talking to yourself. It doesn't matter whether you are connected to the session bus from the Windows side or the WSL side (there isn't really a "side" -- it is a bus). I'm pretty sure you could run the same hacked up `notification-thing` with [dbus-python for Windows](https://mail.gnome.org/archives/meld-list/2012-July/msg00006.html). 

But either way, you are literally firing up a dbus session on every terminal for the privilege of sending a message right back to yourself to spawn New-BurntToastNotification (or your own re-implementation thereof). Which, you can just call directly per yesterday's message, and avoid the drama.

**Comment by therealkenc** [comment]
This puts the "feature request" in an awkward state. Possibilities being:

(1) People just want to get a Win 10 notification, somehow, from the WSL commandline. Which we could do since interop went live.

(2) People _actually want_ a session bus started on every terminal instance. This would live in their `/etc/pam.d/wsld` on Debian derived distros. Which worms fast, and is why #1405 is frozen in time.

(3) This is a coded request for Gnome sessions, because GTK+ is In Real Life the only thing that calls `libnotify`. Which is out of scope.

**Comment by alanaasmaa** [comment]
It would be awesome to have a project that listeners some error logs and send notifications to windows accordingly. 

_It would also be awesome when i would have all my side project finished so i could build this_

**Comment by therealkenc** [comment]
> personally I [would] be happy with the first option. I just want gulp-notify to be able to send me toast notifications.

Yep. Use a [custom notifier](https://www.npmjs.com/package/gulp-notify#notifywithreporterfunction) and call ['powershell.exe -command New-BurntToastNotification'](https://github.com/Microsoft/WSL/issues/2466#issuecomment-370316815) with [`child_process.exec()`](https://nodejs.org/docs/v8.1.4/api/child_process.html#child_process_child_process_exec_command_options_callback).

**Comment by JacobDB** [comment]
@therealkenc oh that's awesome, completely missed that. Thanks, will try to get it integrated in to my system.

**Comment by therealkenc** [comment]
> Thanks, will try to get it integrated in to my system.

Great. 

I think your necropost is probably enough motivation to put this one to bed. The only way this issue can be turned into a WSL actionable is possibility (2). That is highly problematic because WSL is distribution agnostic, and any solution would necessitate overloading WSL's `/init` yet more and firing up a Desktop Notification Specification service _on every bash session_, presumably controlled by some `wsl.conf` setting. On the Windows side `wslhost.exe` (or whatever) would listen on some communication channel (which n.b. is **not** dbus because the dbus session server has to live in each WSL bash session _on the WSL side_). [To help visualize, imagine a Ubuntu WSL instance taking [unity-notification](https://launchpad.net/unity-notifications) messages and a simultaneously running OpenSUSE WSL instance taking our imagined `/init`-to-`wslhost.exe` mechanism.]

Which is just too horrible to contemplate, because having implemented all of that, it gets us nothing option (1) can't do. UserVoice [won't save us](https://github.com/Microsoft/WSL/issues/2466#issuecomment-327275254) here.

**Comment by gravataLonga** [comment]
Any update on this?

**Comment by b-hayes** [comment]
> > personally I [would] be happy with the first option. I just want gulp-notify to be able to send me toast notifications.
> 
> Yep. Use a [custom notifier](https://www.npmjs.com/package/gulp-notify#notifywithreporterfunction) and call ['powershell.exe -command New-BurntToastNotification'](https://github.com/Microsoft/WSL/issues/2466#issuecomment-370316815) with [`child_process.exec()`](https://nodejs.org/docs/v8.1.4/api/child_process.html#child_process_child_process_exec_command_options_callback).

Like the idea of this one but seems there's a lot of hoops to jump through to get it working.
```
New-BurntToastNotification
New-BurntToastNotification : The 'New-BurntToastNotification' command was found in the module 'BurntToast', but the module could not be loaded. For more information, run
'Import-Module BurntToast'.
At line:1 char:1
+ New-BurntToastNotification
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (New-BurntToastNotification:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CouldNotAutoloadMatchingModule

PS C:\Windows\system32> Import-Module BurntToast
Import-Module : File C:\Program Files\WindowsPowerShell\Modules\BurntToast\0.7.1\BurntToast.psm1 cannot be loaded because running scripts is disabled on this system. For more
information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
At line:1 char:1
+ Import-Module BurntToast
+ ~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : SecurityError: (:) [Import-Module], PSSecurityException
    + FullyQualifiedErrorId : UnauthorizedAccess,Microsoft.PowerShell.Commands.ImportModuleCommand
```

**Comment by ykhurshid** [comment]
> 
> 
> > > personally I [would] be happy with the first option. I just want gulp-notify to be able to send me toast notifications.
> > 
> > 
> > Yep. Use a [custom notifier](https://www.npmjs.com/package/gulp-notify#notifywithreporterfunction) and call ['powershell.exe -command New-BurntToastNotification'](https://github.com/Microsoft/WSL/issues/2466#issuecomment-370316815) with [`child_process.exec()`](https://nodejs.org/docs/v8.1.4/api/child_process.html#child_process_child_process_exec_command_options_callback).
> 
> Like the idea of this one but seems there's a lot of hoops to jump through to get it working.
> 
> ```
> New-BurntToastNotification
> New-BurntToastNotification : The 'New-BurntToastNotification' command was found in the module 'BurntToast', but the module could not be loaded. For more information, run
> 'Import-Module BurntToast'.
> At line:1 char:1
> + New-BurntToastNotification
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : ObjectNotFound: (New-BurntToastNotification:String) [], CommandNotFoundException
>     + FullyQualifiedErrorId : CouldNotAutoloadMatchingModule
> 
> PS C:\Windows\system32> Import-Module BurntToast
> Import-Module : File C:\Program Files\WindowsPowerShell\Modules\BurntToast\0.7.1\BurntToast.psm1 cannot be loaded because running scripts is disabled on this system. For more
> information, see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170.
> At line:1 char:1
> + Import-Module BurntToast
> + ~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : SecurityError: (:) [Import-Module], PSSecurityException
>     + FullyQualifiedErrorId : UnauthorizedAccess,Microsoft.PowerShell.Commands.ImportModuleCommand
> ```

You need to set your ExecutionPolicy to unrestricted to make it work. Run Powershell in administrator mode and run the command "Set-ExecutionPolicy -ExecutionPolicy Unrestricted -Scope LocalMachine".

It worked for me after that though the alias was not working for text that ran beyond a single word so I instead made a bash function as follows

` toast () { powershell.exe -command New-BurntToastNotification "-Text '$1'" }`

Stick that into your bashrc (or whatever rc your particular shell happens to use) and you should be good to go.

**Comment by Kallahan23** [comment]
Changing the execution policy directly using Set-ExecutionPolicy works, but leaves your machine open to malicious scripts. I'm not a fan of this approach. I think it's better to leave it set to Restricted.

Instead, I set my alias in WSL to this:

`alias toast='powershell.exe -noprofile -executionpolicy bypass -command New-BurntToastNotification'`

I'm not an expert at Powershell by any means, but as far as I know, this bypasses the execution policy only for this call. That way Powershell can stay restricted by default, but this wsl alias will bypass this and run the toast.

**Comment by WSLUser** [comment]
@therealkenc Since we're getting GUI support for WSL2, I think this issue could be re-opened. This could now be done behind the scenes or it could simply be a configuration option in /etc/wsl.conf if we want Win 10 notifications vice native Linux notifications popping up. I think I'd prefer Win 10 though, easier to manage that way.

**Comment by therealkenc** [comment]
The ask isn't related to GUI support, which was explained in May 2018 ([message](https://github.com/microsoft/WSL/issues/2466#issuecomment-370304415)). That is the case whether X11 or Wayland. You can of course get a _Linux GUI notification_, but that wasn't the ask. The ask was for Windows notifications. Linux GUI notification popups, meanwhile, would be all userspace.

That said, there is no User Voice anymore (there was in 2017), so, sure, re-opening on feature request loophole.

**Comment by xuanruiqi** [comment]
Would it be possible to make an alternate implementation of libnotify based on WSL-Windows interaction mechanisms?

**Comment by bartdorsey** [comment]
I've been trying to get the BurntToast notifier going in WSL.. and I constantly get this error... and I'mnot sure how to fix it.. 
```Exception calling "CreateToastNotifier" with "1" argument(s): "The data area passed to a system call is too small. (Exception from
HRESULT: 0x8007007A)"
At C:\Program Files\WindowsPowerShell\Modules\BurntToast\0.8.5\BurntToast.psm1:2478 char:9
+         [Windows.UI.Notifications.ToastNotificationManager]::CreateTo ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : Exception
```

**Comment by Davikar** [comment]
> I've been trying to get the BurntToast notifier going in WSL.. and I constantly get this error... and I'mnot sure how to fix it..
> 
> ```
> Exception calling "CreateToastNotifier" with "1" argument(s): "The data area passed to a system call is too small. (Exception from
> HRESULT: 0x8007007A)"
> At C:\Program Files\WindowsPowerShell\Modules\BurntToast\0.8.5\BurntToast.psm1:2478 char:9
> +         [Windows.UI.Notifications.ToastNotificationManager]::CreateTo ...
> +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
>     + FullyQualifiedErrorId : Exception
> ```

Having the same issue.

**Comment by b-hayes** [comment]
NGL I can't believe this is still not a thing. Every terminal on every system except windows can do this.

**Comment by kmantel** [comment]
> I've been trying to get the BurntToast notifier going in WSL.. and I constantly get this error... and I'mnot sure how to fix it..
> 
> ```
> Exception calling "CreateToastNotifier" with "1" argument(s): "The data area passed to a system call is too small. (Exception from
> HRESULT: 0x8007007A)"
> At C:\Program Files\WindowsPowerShell\Modules\BurntToast\0.8.5\BurntToast.psm1:2478 char:9
> +         [Windows.UI.Notifications.ToastNotificationManager]::CreateTo ...
> +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
>     + FullyQualifiedErrorId : Exception
> ```



This feels nuts but worked for me 
`powershell powershell.exe -Command New-BurntToastNotification`

**Comment by kmantel** [comment]
> I've been trying to get the BurntToast notifier going in WSL.. and I constantly get this error... and I'mnot sure how to fix it..
> 
> ```
> Exception calling "CreateToastNotifier" with "1" argument(s): "The data area passed to a system call is too small. (Exception from
> HRESULT: 0x8007007A)"
> At C:\Program Files\WindowsPowerShell\Modules\BurntToast\0.8.5\BurntToast.psm1:2478 char:9
> +         [Windows.UI.Notifications.ToastNotificationManager]::CreateTo ...
> +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
>     + FullyQualifiedErrorId : Exception
> ```

This feels silly but worked for me 

`powershell powershell.exe -Command New-BurntToastNotification`

**Comment by shanselman** [comment]
Would be cool @craigloewen-msft

**Comment by ultrabig** [comment]
I created an impelementation using Windows native functionality.
<https://github.com/ultrabig/WslNotifyd>

**Comment by RAraghavarora** [comment]
Boosting this for notifications more than single word!!

> It worked for me after that though the alias was not working for text that ran beyond a single word so I instead made a bash function as follows
> 
> ` toast () { powershell.exe -command New-BurntToastNotification "-Text '$1'" }`
> 
> Stick that into your bashrc (or whatever rc your particular shell happens to use) and you should be good to go.

**Comment by ForbiddenEra** [comment]
> I created an impelementation using Windows native functionality. https://github.com/ultrabig/WslNotifyd

<3 (except not providing a binary.. I don't build windows stuff)

**Comment by petrroll** [comment]
Would be useful for e.g. using herdr (https://herdr.dev/) in WSL
